qtrees/data_processor.py

```diff
@@ -0,0 +1,311 @@
+        relevant_trees = tuple(trees.tree_id.unique())
+        if not self.public_run:
+            trees_private = pd.read_sql("SELECT tree_id,baumscheibe_m2,baumscheibe_surface FROM private.trees_private", self.engine.connect())
+            trees_private["baumscheibe_m2"] = pd.cut(trees_private["baumscheibe_m2"], bins=[0, 5, 100], labels=["klein", "groß"])
+            trees = trees.merge(trees_private, how="left", on="tree_id")
+            if not self.forecast:
+                water = get_watering(relevant_trees)
+                trees = trees.merge(water, how="left", on=["tree_id", "timestamp"])
+        shading = get_shading_index(relevant_trees)
+        trees = trees.merge(shading, how="left", on=["tree_id", "month"])
+        return trees
+
+    def _get_weather_measurements(self):
+        '''Get station data and solar if not public run and merge them. Generate weekly avg or sum depending on column.'''
+        if self.forecast and not self.public_run:
+            weather_station = pd.read_sql("SELECT tile_id, date, wind_max_ms, wind_avg_ms, temp_max_c, temp_avg_c, rainfall_mm FROM private.weather_tile_measurement", con=self.engine.connect())
+            weather_station = weather_station.groupby("date").mean().drop(columns="tile_id").reset_index()
+        else:
+            weather_station = pd.read_sql("SELECT date, wind_max_ms, wind_avg_ms, rainfall_mm, temp_max_c, temp_avg_c, upm FROM public.weather", con=self.engine.connect())
+        weather_list = [weather_station]
+        if not self.public_run:
+            weather_solar = pd.read_sql("SELECT tile_id, date, ghi_sum_whm2 FROM private.weather_tile_measurement", con=self.engine.connect())
+            weather_list.append(weather_solar)
+        for source in weather_list:
+            source["date"] = source["date"].astype('datetime64[ns, UTC]')
+            source.set_index("date", inplace=True)
+            source.interpolate(limit=7)
+        if not self.public_run:
+            weather_station = weather_station.merge(weather_solar.groupby(level=0).mean()["ghi_sum_whm2"], how="left", left_index=True, right_index=True)
+        weather_station["rainfall_mm"] = weather_station["rainfall_mm"].rolling(window=ROLLING_WINDOW, min_periods=1).sum()
+        for col in [x for x in weather_station.columns if x not in ["date", "rainfall_mm"]]:
+            weather_station[col] = weather_station[col].rolling(window=ROLLING_WINDOW, min_periods=1).mean()
+        return weather_station
+
+    def _get_weather_forecast(self):
+        '''For forecast training and inference we take the weather forecast from solar anywhere instead of the station weather. However we do not have upm for the forecast'''
+        weather_solar = pd.read_sql(f"SELECT tile_id, date, ghi_sum_whm2, wind_max_ms, wind_avg_ms, temp_max_c, temp_avg_c, rainfall_mm FROM private.weather_tile_forecast WHERE DATE(created_at) = '{self.date.strftime('%Y-%m-%d')}';", con=self.engine.connect())
```

Comment:
string format

Reply:
In darüberliegender Funktion abgehandelt